File: tlog_lib/tlog.go
package tlog

import (
	"context"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

type Level uint8

const (
	INFO Level = 0
	WARN Level = 1
	ERR  Level = 2
	DBUG Level = 3
)

type ctxKey struct{}

type Context struct {
	TraceID uint64
	SpanID  uint64
	Tags    []string
	Sample  bool
}

var defaultContext = Context{Sample: true}

func GenID() uint64 {
	return rand.Uint64()
}

type Logger struct {
	file       *os.File
	buffer     chan string
	sampleRate float64
	running    atomic.Bool
	wg         sync.WaitGroup
	mu         sync.Mutex
}

var logger *Logger
var once sync.Once

func Get() *Logger {
	once.Do(func() {
		logger = &Logger{
			buffer:     make(chan string, 8192),
			sampleRate: 1.0,
		}
		logger.running.Store(true)
		logger.wg.Add(1)
		go logger.process()
	})
	return logger
}

func (l *Logger) process() {
	defer l.wg.Done()
	for line := range l.buffer {
		l.mu.Lock()
		if l.file != nil {
			l.file.WriteString(line)
		}
		l.mu.Unlock()
	}
}

func (l *Logger) Open(path string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err == nil {
		l.file = f
	}
}

func (l *Logger) Write(ctx context.Context, lvl Level, msg string) {
	tCtx := FromContext(ctx)
	if !tCtx.Sample && lvl != ERR {
		return
	}

	now := time.Now().UnixNano()
	tagsStr := "-"
	if len(tCtx.Tags) > 0 {
		tagsStr = strings.Join(tCtx.Tags, "")
	}

	line := fmt.Sprintf("%016x %016x %016x %d [%s] %s\n",
		now, tCtx.TraceID, tCtx.SpanID, lvl, tagsStr, msg)

	select {
	case l.buffer <- line:
	default:
		// Ring buffer full behavior: drop
	}
}

func FromContext(ctx context.Context) Context {
	if ctx == nil {
		return defaultContext
	}
	if v, ok := ctx.Value(ctxKey{}).(Context); ok {
		return v
	}
	return defaultContext
}

func WithContext(ctx context.Context, tCtx Context) context.Context {
	return context.WithValue(ctx, ctxKey{}, tCtx)
}

// Helper API
func Init(path string) { Get().Open(path) }
func Sample(rate float64) { Get().sampleRate = rate }

func Info(ctx context.Context, msg string) { Get().Write(ctx, INFO, msg) }
func Warn(ctx context.Context, msg string) { Get().Write(ctx, WARN, msg) }
func Err(ctx context.Context, msg string) { Get().Write(ctx, ERR, msg) }

func AddTag(ctx context.Context, k, v string) context.Context {
	tCtx := FromContext(ctx)
	kSafe := strings.ReplaceAll(strings.ReplaceAll(k, " ", "_"), ":", "_")
	vSafe := strings.ReplaceAll(strings.ReplaceAll(v, " ", "_"), ":", "_")
	
	newTags := make([]string, len(tCtx.Tags), len(tCtx.Tags)+1)
	copy(newTags, tCtx.Tags)
	newTags = append(newTags, fmt.Sprintf("%s:%s;", kSafe, vSafe))
	
tCtx.Tags = newTags
	return WithContext(ctx, tCtx)
}

// Span starts a new span scope. Returns the new context and a finish func.
// Usage:
//   ctx, finish := tlog.Span(ctx, "myOp")
//   defer finish()
func Span(ctx context.Context, name string) (context.Context, func()) {
	tCtx := FromContext(ctx)
	
	if tCtx.TraceID == 0 {
		tCtx.TraceID = GenID()
		tCtx.Sample = rand.Float64() <= Get().sampleRate
	}
	tCtx.SpanID = GenID()
	
	newCtx := WithContext(ctx, tCtx)
	Get().Write(newCtx, DBUG, "> "+name)
	
	return newCtx, func() {
		Get().Write(newCtx, DBUG, "< "+name)
	}
}

File: tlog_lib/tlog.lua
local tlog = {}
local unpack = unpack or table.unpack

tlog.Level = { INFO = 0, WARN = 1, ERR = 2, DBUG = 3 }

math.randomseed(math.floor(os.time() + (os.clock() * 1000)))

tlog.ctx = {
    trace_id = nil,
    span_id = nil,
    tags = {},
    sample = true
}

local function gen_id()
    return string.format("%08x%08x", math.random(0, 0x7FFFFFFF), math.random(0, 0x7FFFFFFF))
end

local Logger = {
    file = nil,
    sample_rate = 1.0
}

function Logger:open(path)
    self.file = io.open(path, "a")
end

function Logger:set_sampling(r)
    self.sample_rate = r
end

function Logger:should_sample()
    return math.random() <= self.sample_rate
end

function Logger:write(lvl, msg)
    if not tlog.ctx.sample and lvl ~= tlog.Level.ERR then return end
    
    -- Improvement: Combine os.time (epoch) and os.clock (precision)
    -- This creates a high-resolution hex timestamp the viewer can diff.
    local now_s = os.time()
    local micro = math.floor((os.clock() % 1) * 1000000)
    local timestamp = (now_s * 1000000) + micro

    local tag_str = #tlog.ctx.tags > 0 and table.concat(tlog.ctx.tags, "") or "-"
    local tid = tlog.ctx.trace_id or "0000000000000000"
    local sid = tlog.ctx.span_id or "0000000000000000"
    
    -- Format: timestamp is now %016x (microsecond precision)
    local line = string.format("%016x %s %s %d [%s] %s\n", 
        timestamp, tid, sid, lvl, tag_str, msg)
        
    if self.file then
        self.file:write(line)
        self.file:flush()
    else
        print(line:sub(1, -2))
    end
end

function tlog.init(f) Logger:open(f) end
function tlog.sample(r) Logger:set_sampling(r) end
function tlog.info(m) Logger:write(tlog.Level.INFO, m) end
function tlog.warn(m) Logger:write(tlog.Level.WARN, m) end
function tlog.err(m)  Logger:write(tlog.Level.ERR, m) end

function tlog.tag(k, v)
    local k_s = string.gsub(tostring(k), "[:%s;]", "_")
    local v_s = string.gsub(tostring(v), "[:%s;]", "_")
    table.insert(tlog.ctx.tags, k_s .. ":" .. v_s .. ";")
end

function tlog.scope(name, func, ...)
    local p_tid = tlog.ctx.trace_id
    local p_sid = tlog.ctx.span_id
    local p_smp = tlog.ctx.sample
    local p_tgs = {}
    for i, v in ipairs(tlog.ctx.tags) do p_tgs[i] = v end
    
    if not tlog.ctx.trace_id then
        tlog.ctx.trace_id = gen_id()
        tlog.ctx.sample = Logger:should_sample()
    end
    tlog.ctx.span_id = gen_id()
    
    Logger:write(tlog.Level.DBUG, "> " .. name)
    local res = { pcall(func, ...) }
    Logger:write(tlog.Level.DBUG, "< " .. name)
    
    tlog.ctx.trace_id = p_tid
    tlog.ctx.span_id = p_sid
    tlog.ctx.tags = p_tgs
    tlog.ctx.sample = p_smp
    
    if not res[1] then error(res[2]) end
    return select(2, unpack(res))
end

return tlog
